File: app/src/main/java/com/spreadyourmusic/spreadyourmusic/fragment/PlayerControlsFragment.java
package com.spreadyourmusic.spreadyourmusic.fragment;

import android.app.Fragment;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import com.spreadyourmusic.spreadyourmusic.R;

/**
 * A class that shows the Media Queue to the user.
 */
public class PlayerControlsFragment extends Fragment {

    private ImageButton mPlayPause;
    private TextView mTitle;
    private TextView mSubtitle;
    // Receive callbacks from the MediaController. Here we update our state such as which queue
    // is being shown, the current title and description and the PlaybackState.
    private final MediaControllerCompat.Callback mCallback = new MediaControllerCompat.Callback() {
        @Override
        public void onPlaybackStateChanged(@NonNull PlaybackStateCompat state) {
            PlayerControlsFragment.this.onPlaybackStateChanged(state);
        }

        @Override
        public void onMetadataChanged(MediaMetadataCompat metadata) {
            if (metadata == null) {
                return;
            }
            PlayerControlsFragment.this.onMetadataChanged(metadata);
        }
    };

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_player_controls, container, false);
        mPlayPause = rootView.findViewById(R.id.playPauseSong);
        mPlayPause.setEnabled(true);
        mPlayPause.setOnClickListener(mButtonListener);

        mTitle = rootView.findViewById(R.id.songTitle);
        mSubtitle = rootView.findViewById(R.id.songAuthor);
        return rootView;
    }

    @Override
    public void onStart() {
        super.onStart();
        MediaControllerCompat controller = MediaControllerCompat.getMediaController(getActivity());
        if (controller != null) {
            onConnected();
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        MediaControllerCompat controller = MediaControllerCompat.getMediaController(getActivity());
        if (controller != null) {
            controller.unregisterCallback(mCallback);
        }
    }

    public void onConnected() {
        MediaControllerCompat controller = MediaControllerCompat.getMediaController(getActivity());
        if (controller != null) {
            onMetadataChanged(controller.getMetadata());
            onPlaybackStateChanged(controller.getPlaybackState());
            controller.registerCallback(mCallback);
        }
    }

    private void onMetadataChanged(MediaMetadataCompat metadata) {
        if (getActivity() == null) {
            return;
        }
        if (metadata == null) {
            return;
        }

        mTitle.setText(metadata.getDescription().getTitle());
        mSubtitle.setText(metadata.getDescription().getSubtitle());
    }

    private void onPlaybackStateChanged(PlaybackStateCompat state) {
        if (getActivity() == null) {
            return;
        }
        if (state == null) {
            return;
        }
        boolean enablePlay = false;
        switch (state.getState()) {
            case PlaybackStateCompat.STATE_PAUSED:
            case PlaybackStateCompat.STATE_STOPPED:
                enablePlay = true;
                break;
            case PlaybackStateCompat.STATE_ERROR:
                Toast.makeText(getActivity(), state.getErrorMessage(), Toast.LENGTH_LONG).show();
                break;
        }

        if (enablePlay) {
            mPlayPause.setImageDrawable(
                    ContextCompat.getDrawable(getActivity(), R.drawable.ic_play_circle_filled_white_24dp));
        } else {
            mPlayPause.setImageDrawable(
                    ContextCompat.getDrawable(getActivity(), R.drawable.ic_pause_circle_filled_white_24dp));
        }
    }

    private final View.OnClickListener mButtonListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            MediaControllerCompat controller = MediaControllerCompat.getMediaController(getActivity());
            PlaybackStateCompat stateObj = controller.getPlaybackState();
            final int state = stateObj == null ?
                    PlaybackStateCompat.STATE_NONE : stateObj.getState();
            switch (v.getId()) {
                case R.id.playPauseSong:
                    if (state == PlaybackStateCompat.STATE_PAUSED ||
                            state == PlaybackStateCompat.STATE_STOPPED ||
                            state == PlaybackStateCompat.STATE_NONE) {
                        playMedia();
                    } else if (state == PlaybackStateCompat.STATE_PLAYING ||
                            state == PlaybackStateCompat.STATE_BUFFERING ||
                            state == PlaybackStateCompat.STATE_CONNECTING) {
                        pauseMedia();
                    }
                    break;
            }
        }
    };

    private void playMedia() {
        MediaControllerCompat controller = MediaControllerCompat.getMediaController(getActivity());
        if (controller != null) {
            controller.getTransportControls().play();
        }
    }

    private void pauseMedia() {
        MediaControllerCompat controller = MediaControllerCompat.getMediaController(getActivity());
        if (controller != null) {
            controller.getTransportControls().pause();
        }
    }
}

File: app/src/main/java/com/spreadyourmusic/spreadyourmusic/fragment/PlayerControlsFragment.kt
package com.spreadyourmusic.spreadyourmusic.fragment

import android.content.Intent
import android.support.v4.app.Fragment
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import android.widget.Toast

import com.spreadyourmusic.spreadyourmusic.R
import com.spreadyourmusic.spreadyourmusic.activities.PlayerActivity
import com.spreadyourmusic.spreadyourmusic.controller.getCurrentSong
import com.spreadyourmusic.spreadyourmusic.media.playback.MusicQueueManager

/**
 * Player buttons controller
 */
class PlayerControlsFragment : Fragment() {

    private var mPlayPause: ImageButton? = null
    private var mTitle: TextView? = null
    private var mSubtitle: TextView? = null
    private var nextSongImageButton:ImageButton? = null
    private var previousSongImageButton: ImageButton? = null

    private val mCallback = object : MediaControllerCompat.Callback() {
        override fun onPlaybackStateChanged(state: PlaybackStateCompat) {
            this@PlayerControlsFragment.onPlaybackStateChanged(state)
        }

        override fun onMetadataChanged(metadata: MediaMetadataCompat?) {
            if (metadata == null) {
                return
            }
            this@PlayerControlsFragment.onMetadataChanged(metadata)
        }
    }

    private val mPlayPauseListener = View.OnClickListener { v ->
        val controller = MediaControllerCompat.getMediaController(activity!!)
        val stateObj = controller.playbackState
        val state = stateObj?.state ?: PlaybackStateCompat.STATE_NONE
        when (v.id) {
            R.id.playPauseSong -> if (state == PlaybackStateCompat.STATE_PAUSED ||
                    state == PlaybackStateCompat.STATE_STOPPED ||
                    state == PlaybackStateCompat.STATE_NONE) {
                playMedia()
            } else if (state == PlaybackStateCompat.STATE_PLAYING ||
                    state == PlaybackStateCompat.STATE_BUFFERING ||
                    state == PlaybackStateCompat.STATE_CONNECTING) {
                pauseMedia()
            }
        }
    }

    private val mNextSongListener = View.OnClickListener { _ ->
        val controls = MediaControllerCompat.getMediaController(activity!!).transportControls
        controls.skipToNext()
    }

    private val mPreviousSongListener = View.OnClickListener { _ ->
        val controls = MediaControllerCompat.getMediaController(activity!!).transportControls
        controls.skipToPrevious()
    }

    private val mBackgroundListener = View.OnClickListener { _ ->
        val intent = Intent(activity, PlayerActivity::class.java)
        startActivity(intent)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_player_controls, container, false)
        mPlayPause = rootView.findViewById(R.id.playPauseSong)
        mPlayPause!!.isEnabled = true
        mPlayPause!!.setOnClickListener(mPlayPauseListener)

        mTitle = rootView.findViewById(R.id.songTitle)
        mSubtitle = rootView.findViewById(R.id.songAuthor)

        nextSongImageButton = rootView.findViewById(R.id.nextSong)
        nextSongImageButton!!.setOnClickListener(mNextSongListener)
        previousSongImageButton = rootView.findViewById(R.id.previousSong)
        previousSongImageButton!!.setOnClickListener(mPreviousSongListener)

        val backgroundLayout : View = rootView.findViewById(R.id.fondo)
        backgroundLayout.setOnClickListener(mBackgroundListener)
        mTitle!!.setOnClickListener(mBackgroundListener)
        mSubtitle!!.setOnClickListener(mBackgroundListener)

        return rootView
    }

    override fun onStart() {
        super.onStart()
        val controller = MediaControllerCompat.getMediaController(activity!!)
        if (controller != null) {
            onConnected()
        }
    }

    override fun onStop() {
        super.onStop()
        val controller = MediaControllerCompat.getMediaController(activity!!)
        controller?.unregisterCallback(mCallback)
    }

    fun onConnected() {
        val controller = MediaControllerCompat.getMediaController(activity!!)
        if (controller != null) {
            onMetadataChanged(controller.metadata)
            onPlaybackStateChanged(controller.playbackState)
            controller.registerCallback(mCallback)
        }
    }

    private fun onMetadataChanged(metadata: MediaMetadataCompat?) {
        if (activity != null && metadata != null) {
            val currentSong = getCurrentSong()
            mTitle!!.text = currentSong.name
            mSubtitle!!.text = currentSong.album.creator.username
            nextSongImageButton!!.visibility = if (MusicQueueManager.getInstance().currentQueueSize == 1)
                View.INVISIBLE
            else
                View.VISIBLE
            previousSongImageButton!!.visibility = nextSongImageButton!!.visibility
        }
    }

    private fun onPlaybackStateChanged(state: PlaybackStateCompat?) {
        if (activity != null && state != null) {
            var enablePlay = false
            when (state.state) {
                PlaybackStateCompat.STATE_PAUSED, PlaybackStateCompat.STATE_STOPPED -> enablePlay = true
                PlaybackStateCompat.STATE_ERROR -> Toast.makeText(activity, state.errorMessage, Toast.LENGTH_LONG).show()
            }

            if (enablePlay) {
                mPlayPause!!.setImageDrawable(
                        ContextCompat.getDrawable(activity!!, R.drawable.ic_play_circle_filled_white_24dp))
            } else {
                mPlayPause!!.setImageDrawable(
                        ContextCompat.getDrawable(activity!!, R.drawable.ic_pause_circle_filled_white_24dp))
            }
        }
    }

    private fun playMedia() {
        val controller = MediaControllerCompat.getMediaController(activity!!)
        controller?.transportControls?.play()
    }

    private fun pauseMedia() {
        val controller = MediaControllerCompat.getMediaController(activity!!)
        controller?.transportControls?.pause()
    }
}
